### PR TITLE
Video view with details (155948955)

### PR DESCRIPTION
**Pull request by CodingCarlos**
Added view, with collapsed options, and user view.

Need to add:
 - [x] Default user picture

### CODE REVIEW

File: pages/video-detail/video-detail.view.scss [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		max-width: 1200px;
+		@extend %display-flex;
+		@extend %flex-direction-column;
+		@extend %justify-content-space-between;
+		@extend %align-items-center;
+		@include desktop {
+			margin: 20px auto 0 auto;
+			width: 96vw;
+			background-color: white;
+			flex-direction: row;
```
Comment (DevStarlight) [inline pages/video-detail/video-detail.view.scss]: hmm compatibility here =)

**Review by DevStarlight — APPROVED** [review]

**Comment by CodingCarlos** [comment]
Updated with qa feedback

File: filters/locale-date.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+(function () {
```
Comment (javierLiarte) [inline filters/locale-date.js:1]: I like the filters approach, storing raw data in backend and transforming as needed when it has to be displayed  :+1: 

I gess they're used as translate in view templates.

If we just decouple the function from the angular filter definition, I suppose that this also allow us to reuse the functions code if needed, packaged in a node module or similar, to use the same code to transform in frontend and in backend if some time we need it there also...

File: languages/es_es.js [hunk 3]
```diff
@@ -49,22 +49,47 @@
...
+			
+			VIDEO_ERROR: 'Vídeo erróneo',
+			VIDEO_NOT_FOUND: 'No hemos podido encontrar ese vídeo. Por favor, comprueba que el enlace es correcto, y prueba de nuevo.',
+
+			DROP_VIDEO: 'Arrastra tu vídeo o haz click para subirlo',
+			UPLOADING_FILE: 'Subiendo tu vídeo...',
 			UPLOAD_TITLE_PLACEHOLDER: 'Título',
 			UPLOAD_DESCRIPTION_PLACEHOLDER: 'Descripcion',
-			UPLOAD_SEND: 'Subir video',
+			UPLOAD_SEND: 'Subir vídeo',
```
Comment (javierLiarte) [inline languages/es_es.js:81]: check str

File: pages/video-detail/video-detail.controller.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	self.video = video;
+	self.user = user;
+
+	self.showMore = showMore();
+
+	if(self.video && self.video.data && self.video.data.id !== self.id) {
+		self.video.reset();
+	}
+	
+	// ToDo: Check invalid video :S
```
Comment (javierLiarte) [inline pages/video-detail/video-detail.controller.js:22]: we should have a 404 directive, template or redirect to not get lazy with this error checks in the begining. If later on we want to treat error more specifically, we can override behaviour, but this way, we just use the default error handler.

File: pages/video-detail/video-detail.controller.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	self.video.get(self.id, function() {
+		self.loading = false;
+		self.user.get(self.video.data.owner, function() {
```
Comment (javierLiarte) [inline pages/video-detail/video-detail.controller.js:25]: This hasn't to be tackled now, but from the experience of this cycle working with frontend, for me this kind of pattern of use has some weird smell. I'll explain myself:

* If services are singletons, then they are going to persist in subsequent calls, so they are their data.
* This lead to extrange or not desired behaviours such as one than happened in video donwload page, that showed data from last visited video instead of current one, and extrange workarrounds like videoService.reset()
* It also add implicit dependency of view to service, as view access to Controller.service.data, and if I change the implementation of my service, I would need to change both controller and view
* For me a service would allow me to retrieve some data, but I will store and manage in the controller, passing it back to service if further modifications are needed (such as entity update). Thus, the service will provide me with a repository pattern for the chosen model, and abstracts the controller in how to get and manage the data (as it's done in backend)
* Another use case is to store locally data that should not change, such as (maybe) the curren logged in user. In this case I'll maybe choose a singleton to store (and eventually update) it.

In summary, it makes more sense to me, not to store this kind of data on services singlentons, but instead on controller instances. Controller retrieves somehow (from the service in this case, from local caches, maybe in others) an object and stores locally, and view just know that controller has an object to display.

File: pages/video-detail/video-detail.view.html [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	
+	<article ng-if="Video.video.data && !Video.video.data.error">
+		<section>
+			<div>
+				<video controls ng-src="{{::Video.video.data.video}}" ng-attr-poster="{{::Video.video.data.poster}}" preload="metadata" controlsList="nodownload"></video>
+				<!-- Only if desktop >>> -->
+				<section class="card-user-info card-user-info-desktop">
+					<div>
+						<img ng-if="Video.user.data.pic" ng-src="{{Video.user.data.pic}}">
+						<ng-include ng-if="!Video.user.data.pic" src="'/img/no-user.svg'" ui-sref="userGallery({userId: Video.user.data.owner})"></ng-include>
```
Comment (javierLiarte) [inline pages/video-detail/video-detail.view.html]: great! param fixed now! :+1:

**Review by javierLiarte — COMMENTED** [review]
Great job, guys! Just one video string and one layered architecture related comment

**Comment by javierLiarte** [comment]
Also make sure of update versioning before merging, that I almost forget :)

File: filters/locale-date.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	angular.module('app')
+		.filter('localeDate', ['$translate', localeDate]);
```
Comment (DevStarlight) [inline filters/locale-date.js]: wouldn't it easier to apply the predefined [date](https://docs.angularjs.org/api/ng/filter/date) filter?

File: filters/locale-date.js (hunk 2 above)
Comment (DevStarlight) [reply filters/locale-date.js:1]: wouldn't it easier to apply the predefined [date](https://docs.angularjs.org/api/ng/filter/date) filter? @CodingCarlos @javierLiarte

File: filters/time.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+(function () {
```
Comment (DevStarlight) [inline filters/time.js:1]: wouldn't it easier to apply the predefined date filter? @CodingCarlos @javierLiarte

**Review by DevStarlight — CHANGES_REQUESTED** [review]
until discussing about what I left as a feedback, I'm requesting changes in this PR

File: filters/locale-date.js (hunk 2 above)
Comment (CodingCarlos) [reply filters/locale-date.js:1]: Are we needing this in other place? If so, for sure we shall create a package. If not, maybe it's not the moment to do that (?)

Comment (CodingCarlos) [reply filters/locale-date.js:1]: @DevStarlight Date filter doesn't allow you to set the date in design required format. Or if it does, I didn't found the way. Specially, to use locale strings.

File: filters/time.js (hunk 8 above)
Comment (CodingCarlos) [reply filters/time.js:1]: @DevStarlight As far as I know, there's no default time filter, but just the [date filter](https://docs.angularjs.org/api/ng/filter/date), and it does not convert seconds into `mm:ss.sss` format, but shows in that format a date's moment ;) 

This makes me think in adding jsdocs to the filters, and maybe to all the code.

File: pages/video-detail/video-detail.controller.js (hunk 5 above)
Comment (CodingCarlos) [reply pages/video-detail/video-detail.controller.js:25]: Humm... There's a big topic here, and actually, it depends on so much things. 
 - If you move the data from service to controller, you can't share that data through different controllers, at least in a recommended way. 
 - Also, if you move it to controller, yes, you're fit to service definitions, but you only define the way to manage data in one point of the code instead of doing it several times, one per controller.
 - Keeping the data in services allows you to automatically pre-cache results for several requests (since controllers dies when you get out the view, so it don't keep any previous request)

We can set a day, and do together (cc @DevStarlight) some workarrounds and decide. Might be the same day we talk about async and Ecma rules ;)

File: pages/video-detail/video-detail.controller.js (hunk 4 above)
Comment (CodingCarlos) [reply pages/video-detail/video-detail.controller.js:22]: Definitely yes. But, when talking with @PabloFMM and Juan, It seemed better to have different 404 implementations for each type (ie, "Video not exist", "User not exist"...)

File: filters/locale-date.js (hunk 2 above)
Comment (javierLiarte) [reply filters/locale-date.js:1]: No, i think it's not needed by now, just thinking about it :)

And for the date localization, it seems it can be done with an angular module, if we want to explore it for the future: 
https://stackoverflow.com/a/32111231

File: pages/video-detail/video-detail.controller.js (hunk 4 above)
Comment (javierLiarte) [reply pages/video-detail/video-detail.controller.js:22]: I gues that could be also some kind of parameter of the template/directive, etc

File: pages/video-detail/video-detail.controller.js (hunk 5 above)
Comment (javierLiarte) [reply pages/video-detail/video-detail.controller.js:25]: yep, definitively :)

For me it's all about the two use cases that I mentioned avobe:
* Generally you don't want to share data through different controllers, avoiding extrange behaviour mentioned.
* Sometimes you want, as currently logged in user, then it has sense to have that data in a singleton ;)

**Comment by javierLiarte** [comment]
For me, new changes here seems correct!

**Comment by DevStarlight** [comment]
After talking to @CodingCarlos, the reason for what we're using a custom filter instead of the default one is just simply it doesn't cover all cases we need. After that talk, for me it's fine and we can merge this branch.

**Review by DevStarlight — APPROVED** [review]
